File: ts/src/integrations/openai/proxy.ts
/**
 * ClientProxy — Proxy-based wrapper around an OpenAI client.
 *
 * Intercepts .chat.completions.create / .responses.create. All other
 * attribute access passes through transparently. Spec §4.1 + §6.2.
 * Mirror of Python ``_proxy.py``.
 */
import { ulid } from "ulid";
import type { TraceSink } from "./sink.js";
import { currentSession } from "./session.js";
import { mapExceptionToReason } from "./taxonomy.js";
import {
  buildRequestSnapshot,
  buildSuccessTrace,
  buildFailureTrace,
  finalizeStreamingTrace,
  type RequestSnapshot,
} from "./trace-builder.js";
import {
  buildProviderSourceInfo,
  finishInvocationTiming,
  resolveProviderIdentity,
  startInvocationClock,
} from "../_shared/proxy-runtime.js";
import { AsyncStreamProxy } from "./stream-proxy.js";

export const WRAPPED_SENTINEL = Symbol.for("autocontext.wrapped");

function _isAsyncClient(client: unknown): boolean {
  // Check by class name to avoid ESM require() issues with the openai package
  const className = (client as object)?.constructor?.name ?? "";
  return className.startsWith("Async");
}

export class ClientProxy {
  private readonly _inner: unknown;
  private readonly _sink: TraceSink;
  private readonly _appId: string;
  private readonly _environmentTag: string;
  private readonly _isAsync: boolean;

  constructor(opts: {
    inner: unknown;
    sink: TraceSink;
    appId: string;
    environmentTag: string;
  }) {
    this._inner = opts.inner;
    this._sink = opts.sink;
    this._appId = opts.appId;
    this._environmentTag = opts.environmentTag;
    this._isAsync = _isAsyncClient(opts.inner);
  }

  _sourceInfo(): { emitter: string; sdk: { name: string; version: string } } {
    return buildProviderSourceInfo(import.meta.url);
  }

  _env(): { environmentTag: string; appId: string } {
    return { environmentTag: this._environmentTag, appId: this._appId };
  }

  _invokeChatCompletionsCreate(kwargs: Record<string, unknown>): unknown {
    if (kwargs["stream"]) {
      if (this._isAsync) return this._invokeStreamingAsync(kwargs);
      return this._invokeStreaming(kwargs);
    }
    if (this._isAsync) return this._invokeNonStreamingAsync(kwargs);
    return this._invokeNonStreaming(kwargs);
  }

  _invokeNonStreaming(kwargs: Record<string, unknown>): unknown {
    const perCall = kwargs["autocontext"] as Record<string, string> | null;
    delete kwargs["autocontext"];
    const identity = resolveProviderIdentity(perCall, currentSession());
    const snapshot = buildRequestSnapshot({
      model: String(kwargs["model"] ?? ""),
      messages: (kwargs["messages"] as Array<Record<string, unknown>>) ?? [],
      extraKwargs: Object.fromEntries(
        Object.entries(kwargs).filter(([k]) => k !== "model" && k !== "messages"),
      ),
    });
    const clock = startInvocationClock();
    let response: unknown;
    try {
      const inner = this._inner as Record<string, { completions: { create: (k: unknown) => unknown } }>;
      response = inner["chat"]["completions"]["create"](kwargs);
    } catch (exc) {
      const timing = finishInvocationTiming(clock);
      const trace = buildFailureTrace({
        requestSnapshot: snapshot,
        identity,
        timing,
        env: this._env(),
        sourceInfo: this._sourceInfo(),
        traceId: ulid(),
        reasonKey: mapExceptionToReason(exc),
        errorMessage: String(exc),
        stack: exc instanceof Error ? (exc.stack ?? null) : null,
      });
      this._sink.add(trace as unknown as Record<string, unknown>);
      throw exc;
    }
    // Response is a Promise for async, but for sync OpenAI this is direct
    return (response as Promise<unknown>).then(
      (resp) => {
        const timing = finishInvocationTiming(clock);
        const r = resp as Record<string, unknown>;
        const usage = r["usage"] as Record<string, unknown> | null;
        let toolCalls: Array<Record<string, unknown>> | null = null;
        const choices = r["choices"] as Array<Record<string, unknown>> | undefined;
        if (choices && choices.length > 0) {
          const msg = (choices[0]!["message"] as Record<string, unknown>);
          const tcs = msg?.["tool_calls"] as Array<Record<string, unknown>> | null;
          if (tcs && tcs.length > 0) toolCalls = tcs;
        }
        const trace = buildSuccessTrace({
          requestSnapshot: snapshot,
          responseUsage: usage,
          responseToolCalls: toolCalls,
          identity,
          timing,
          env: this._env(),
          sourceInfo: this._sourceInfo(),
          traceId: ulid(),
        });
        this._sink.add(trace as unknown as Record<string, unknown>);
        return resp;
      },
      (exc: unknown) => {
        const timing = finishInvocationTiming(clock);
        const trace = buildFailureTrace({
          requestSnapshot: snapshot,
          identity,
          timing,
          env: this._env(),
          sourceInfo: this._sourceInfo(),
          traceId: ulid(),
          reasonKey: mapExceptionToReason(exc),
          errorMessage: String(exc),
          stack: exc instanceof Error ? (exc.stack ?? null) : null,
        });
        this._sink.add(trace as unknown as Record<string, unknown>);
        throw exc;
      },
    );
  }

  async _invokeNonStreamingAsync(kwargs: Record<string, unknown>): Promise<unknown> {
    const perCall = kwargs["autocontext"] as Record<string, string> | null;
    delete kwargs["autocontext"];
    const identity = resolveProviderIdentity(perCall, currentSession());
    const snapshot = buildRequestSnapshot({
      model: String(kwargs["model"] ?? ""),
      messages: (kwargs["messages"] as Array<Record<string, unknown>>) ?? [],
      extraKwargs: Object.fromEntries(
        Object.entries(kwargs).filter(([k]) => k !== "model" && k !== "messages"),
      ),
    });
    const clock = startInvocationClock();
    let resp: unknown;
    try {
      const inner = this._inner as Record<string, { completions: { create: (k: unknown) => Promise<unknown> } }>;
      resp = await inner["chat"]["completions"]["create"](kwargs);
    } catch (exc) {
      const timing = finishInvocationTiming(clock);
      const trace = buildFailureTrace({
        requestSnapshot: snapshot,
        identity,
        timing,
        env: this._env(),
        sourceInfo: this._sourceInfo(),
        traceId: ulid(),
        reasonKey: mapExceptionToReason(exc),
        errorMessage: String(exc),
        stack: exc instanceof Error ? (exc.stack ?? null) : null,
      });
      this._sink.add(trace as unknown as Record<string, unknown>);
      throw exc;
    }
    const timing = finishInvocationTiming(clock);
    const r = resp as Record<string, unknown>;
    const usage = r["usage"] as Record<string, unknown> | null;
    let toolCalls: Array<Record<string, unknown>> | null = null;
    const choices = r["choices"] as Array<Record<string, unknown>> | undefined;
    if (choices && choices.length > 0) {
      const msg = choices[0]!["message"] as Record<string, unknown>;
      const tcs = msg?.["tool_calls"] as Array<Record<string, unknown>> | null;
      if (tcs && tcs.length > 0) toolCalls = tcs;
    }
    const trace = buildSuccessTrace({
      requestSnapshot: snapshot,
      responseUsage: usage,
      responseToolCalls: toolCalls,
      identity,
      timing,
      env: this._env(),
      sourceInfo: this._sourceInfo(),
      traceId: ulid(),
    });
    this._sink.add(trace as unknown as Record<string, unknown>);
    return resp;
  }

  _invokeStreaming(kwargs: Record<string, unknown>): unknown {
    const perCall = kwargs["autocontext"] as Record<string, string> | null;
    delete kwargs["autocontext"];
    // Auto-inject stream_options.include_usage = true if absent (not set by caller)
    const streamOpts = Object.assign({}, (kwargs["stream_options"] as Record<string, unknown>) ?? {});
    if (!("include_usage" in streamOpts)) {
      streamOpts["include_usage"] = true;
      kwargs["stream_options"] = streamOpts;
    }
    const identity = resolveProviderIdentity(perCall, currentSession());
    const snapshot = buildRequestSnapshot({
      model: String(kwargs["model"] ?? ""),
      messages: (kwargs["messages"] as Array<Record<string, unknown>>) ?? [],
      extraKwargs: Object.fromEntries(
        Object.entries(kwargs).filter(([k]) => k !== "model" && k !== "messages"),
      ),
    });
    const clock = startInvocationClock();
    const sink = this._sink;
    const env = this._env();
    const sourceInfo = this._sourceInfo();

    const inner = this._inner as Record<string, { completions: { create: (k: unknown) => unknown } }>;
    const streamResult = inner["chat"]["completions"]["create"](kwargs);

    // acc_ref avoids a cycle: proxy → on_finalize → acc_ref → proxy._accumulator
    const accRef: { accumulator: Record<string, unknown> | null } = { accumulator: null };

    const onFinalize = (outcome: Record<string, unknown>): void => {
      const timing = finishInvocationTiming(clock);
      const acc = accRef.accumulator ?? { usage: null, toolCalls: null };
      const trace = finalizeStreamingTrace({
        requestSnapshot: snapshot,
        identity,
        timing,
        env,
        sourceInfo,
        traceId: ulid(),
        accumulatedUsage: (acc["usage"] as Record<string, unknown>) ?? null,
        accumulatedToolCalls: (acc["tool_calls"] as Array<Record<string, unknown>>) ?? null,
        outcome,
      });
      sink.add(trace as unknown as Record<string, unknown>);
    };

    // The proxy wraps the stream promise
    const proxy = new AsyncStreamProxy({ innerStream: streamResult, onFinalize });
    accRef.accumulator = proxy._accumulator;
    return proxy;
  }

  async _invokeStreamingAsync(kwargs: Record<string, unknown>): Promise<unknown> {
    const perCall = kwargs["autocontext"] as Record<string, string> | null;
    delete kwargs["autocontext"];
    // Auto-inject stream_options.include_usage = true if absent
    const streamOpts = Object.assign({}, (kwargs["stream_options"] as Record<string, unknown>) ?? {});
    if (!("include_usage" in streamOpts)) {
      streamOpts["include_usage"] = true;
      kwargs["stream_options"] = streamOpts;
    }
    const identity = resolveProviderIdentity(perCall, currentSession());
    const snapshot = buildRequestSnapshot({
      model: String(kwargs["model"] ?? ""),
      messages: (kwargs["messages"] as Array<Record<string, unknown>>) ?? [],
      extraKwargs: Object.fromEntries(
        Object.entries(kwargs).filter(([k]) => k !== "model" && k !== "messages"),
      ),
    });
    const clock = startInvocationClock();
    const sink = this._sink;
    const env = this._env();
    const sourceInfo = this._sourceInfo();

    const accRef: { accumulator: Record<string, unknown> | null } = { accumulator: null };

    const onFinalize = (outcome: Record<string, unknown>): void => {
      const timing = finishInvocationTiming(clock);
      const acc = accRef.accumulator ?? { usage: null, tool_calls: null };
      const trace = finalizeStreamingTrace({
        requestSnapshot: snapshot,
        identity,
        timing,
        env,
        sourceInfo,
        traceId: ulid(),
        accumulatedUsage: (acc["usage"] as Record<string, unknown>) ?? null,
        accumulatedToolCalls: (acc["tool_calls"] as Array<Record<string, unknown>>) ?? null,
        outcome,
      });
      sink.add(trace as unknown as Record<string, unknown>);
    };

    const inner = this._inner as Record<string, { completions: { create: (k: unknown) => Promise<unknown> } }>;
    const rawStream: unknown = inner["chat"]["completions"]["create"](kwargs);
    let innerStream: unknown = rawStream;
    if (rawStream && typeof (rawStream as { then?: unknown }).then === "function") {
      innerStream = await (rawStream as Promise<unknown>);
    }

    const proxy = new AsyncStreamProxy({ innerStream, onFinalize });
    accRef.accumulator = proxy._accumulator;
    return proxy;
  }

  _invokeResponsesCreate(
    kwargs: Record<string, unknown>,
    normalizedMessages: Array<Record<string, unknown>>,
  ): unknown {
    const perCall = kwargs["autocontext"] as Record<string, string> | null;
    delete kwargs["autocontext"];
    const identity = resolveProviderIdentity(perCall, currentSession());
    const model = String(kwargs["model"] ?? "");
    const snapshot = buildRequestSnapshot({
      model,
      messages: normalizedMessages,
      extraKwargs: Object.fromEntries(
        Object.entries(kwargs).filter(([k]) => k !== "model" && k !== "messages" && k !== "input"),
      ),
    });
    const clock = startInvocationClock();
    const inner = this._inner as Record<string, { create: (k: unknown) => unknown }>;
    const result = inner["responses"]["create"](kwargs);
    return (result as Promise<unknown>).then(
      (resp) => {
        const timing = finishInvocationTiming(clock);
        const r = resp as Record<string, unknown>;
        const usage = r["usage"] as Record<string, unknown> | null;
        const trace = buildSuccessTrace({
          requestSnapshot: snapshot,
          responseUsage: usage,
          responseToolCalls: null,
          identity,
          timing,
          env: this._env(),
          sourceInfo: this._sourceInfo(),
          traceId: ulid(),
        });
        this._sink.add(trace as unknown as Record<string, unknown>);
        return resp;
      },
      (exc: unknown) => {
        const timing = finishInvocationTiming(clock);
        const trace = buildFailureTrace({
          requestSnapshot: snapshot,
          identity,
          timing,
          env: this._env(),
          sourceInfo: this._sourceInfo(),
          traceId: ulid(),
          reasonKey: mapExceptionToReason(exc),
          errorMessage: String(exc),
          stack: exc instanceof Error ? (exc.stack ?? null) : null,
        });
        this._sink.add(trace as unknown as Record<string, unknown>);
        throw exc;
      },
    );
  }
}

File: autocontext/src/autocontext/integrations/openai/_proxy.py
"""ClientProxy — attribute-delegating wrapper around an OpenAI client.

Intercepts ``.chat.completions.create`` / ``.chat.completions.create``-async /
``.responses.create`` / ``.responses.create``-async. All other attribute
access passes through transparently. Spec §4.1 + §6.2.

Streaming is handled by ``_stream.StreamProxy`` — this module only dispatches.
"""
from __future__ import annotations

import inspect
import time
import traceback
from datetime import UTC, datetime
from typing import Any

from ulid import ULID

from autocontext.integrations._shared.identity import resolve_identity
from autocontext.integrations.openai._sink import TraceSink
from autocontext.integrations.openai._taxonomy import map_exception_to_reason
from autocontext.integrations.openai._trace_builder import (
    build_failure_trace,
    build_request_snapshot,
    build_success_trace,
)


def _is_async_client(client: Any) -> bool:
    """Return True if client is an AsyncOpenAI (or compatible async client)."""
    try:
        from openai import AsyncOpenAI  # noqa: PLC0415
        return isinstance(client, AsyncOpenAI)
    except ImportError:
        pass
    # Fallback: check class name
    return type(client).__name__.startswith("Async")

_WRAPPED_SENTINEL = "__autocontext_wrapped__"


def _now_iso() -> str:
    return datetime.now(UTC).isoformat().replace("+00:00", "Z")


class _ChatCompletionsProxy:
    def __init__(self, parent: ClientProxy, inner_create: Any) -> None:
        self._parent = parent
        self._inner_create = inner_create

    def create(self, **kwargs: Any) -> Any:
        if kwargs.get("stream", False):
            if self._parent._is_async:
                return self._parent._invoke_streaming_async(
                    inner_method=self._inner_create, kwargs=kwargs
                )
            return self._parent._invoke_streaming(
                inner_method=self._inner_create, kwargs=kwargs
            )
        if self._parent._is_async:
            return self._parent._invoke_non_streaming_async(
                inner_method=self._inner_create, kwargs=kwargs,
            )
        return self._parent._invoke_non_streaming(
            inner_method=self._inner_create, kwargs=kwargs,
        )


class _ChatProxy:
    def __init__(self, parent: ClientProxy, inner_chat: Any) -> None:
        self._parent = parent
        self._inner_chat = inner_chat

    @property
    def completions(self) -> _ChatCompletionsProxy:
        return _ChatCompletionsProxy(self._parent, self._inner_chat.completions.create)


class _ResponsesProxy:
    def __init__(self, parent: ClientProxy, inner_responses_create: Any) -> None:
        self._parent = parent
        self._inner_create = inner_responses_create

    def create(self, **kwargs: Any) -> Any:
        # responses.create currently shares the same request/response envelope
        # semantics vs chat.completions — messages come in as `input` (single
        # string or list of content blocks). For v1 coverage we pass the input
        # through as-is under `messages` key in the trace for schema compatibility.
        normalized_messages = kwargs.get("messages") or [
            {"role": "user", "content": kwargs.get("input", "")}
        ]
        kwargs_for_trace = dict(kwargs)
        kwargs_for_trace["messages"] = normalized_messages
        kwargs_for_trace.pop("input", None)
        if kwargs.get("stream", False):
            if self._parent._is_async:
                raise NotImplementedError("async responses streaming — see deferred list")
            return self._parent._invoke_streaming(
                inner_method=self._inner_create, kwargs=kwargs,
            )
        if self._parent._is_async:
            return self._parent._invoke_non_streaming_async_responses(
                inner_method=self._inner_create, kwargs=kwargs,
                normalized_messages=normalized_messages,
            )
        return self._parent._invoke_non_streaming_responses(
            inner_method=self._inner_create, kwargs=kwargs,
            normalized_messages=normalized_messages,
        )


class ClientProxy:
    def __init__(
        self,
        *,
        inner: Any,
        sink: TraceSink,
        app_id: str,
        environment_tag: str,
    ) -> None:
        object.__setattr__(self, "_inner", inner)
        object.__setattr__(self, "_sink", sink)
        object.__setattr__(self, "_app_id", app_id)
        object.__setattr__(self, "_environment_tag", environment_tag)
        object.__setattr__(self, "_is_async", _is_async_client(inner))
        object.__setattr__(self, _WRAPPED_SENTINEL, True)

    def __getattr__(self, name: str) -> Any:
        if name == "chat":
            return _ChatProxy(self, self._inner.chat)
        if name == "responses":
            return _ResponsesProxy(self, self._inner.responses.create)
        return getattr(self._inner, name)

    def _source_info(self) -> dict[str, Any]:
        try:
            from importlib.metadata import version
            ver = version("autocontext")
        except Exception:
            ver = "0.0.0"
        return {"emitter": "sdk", "sdk": {"name": "autocontext-py", "version": ver}}

    def _env(self) -> dict[str, Any]:
        return {"environmentTag": self._environment_tag, "appId": self._app_id}

    def _invoke_non_streaming(
        self,
        *,
        inner_method: Any,
        kwargs: dict[str, Any],
    ) -> Any:
        per_call = kwargs.pop("autocontext", None)
        if kwargs.get("stream", False):
            raise NotImplementedError("streaming not yet wired")
        identity = resolve_identity(per_call)
        request_snapshot = build_request_snapshot(
            model=kwargs.get("model", ""),
            messages=kwargs.get("messages", []),
            extra_kwargs={k: v for k, v in kwargs.items() if k not in {"model", "messages"}},
        )
        started_at = _now_iso()
        started_monotonic = time.monotonic()
        try:
            response = inner_method(**kwargs)
        except Exception as exc:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            trace = build_failure_trace(
                request_snapshot=request_snapshot,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=self._env(),
                source_info=self._source_info(),
                trace_id=str(ULID()),
                reason_key=map_exception_to_reason(exc),
                error_message=str(exc),
                stack=traceback.format_exc(),
            )
            self._sink.add(trace)
            raise
        ended_at = _now_iso()
        latency_ms = int((time.monotonic() - started_monotonic) * 1000)
        usage = response.usage.model_dump() if getattr(response, "usage", None) else None
        tool_calls = None
        if hasattr(response, "choices") and response.choices and response.choices[0].message.tool_calls:
            tool_calls = [tc.model_dump() for tc in response.choices[0].message.tool_calls]
        trace = build_success_trace(
            request_snapshot=request_snapshot,
            response_usage=usage,
            response_tool_calls=tool_calls,
            identity=identity,
            timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
            env=self._env(),
            source_info=self._source_info(),
            trace_id=str(ULID()),
        )
        self._sink.add(trace)
        return response

    def _invoke_non_streaming_responses(
        self,
        *,
        inner_method: Any,
        kwargs: dict[str, Any],
        normalized_messages: list[dict[str, Any]],
    ) -> Any:
        per_call = kwargs.pop("autocontext", None)
        identity = resolve_identity(per_call)
        model = kwargs.get("model", "")
        request_snapshot = build_request_snapshot(
            model=model,
            messages=normalized_messages,
            extra_kwargs={k: v for k, v in kwargs.items() if k not in {"model", "messages", "input"}},
        )
        started_at = _now_iso()
        started_monotonic = time.monotonic()
        try:
            response = inner_method(**kwargs)
        except Exception as exc:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            trace = build_failure_trace(
                request_snapshot=request_snapshot,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=self._env(),
                source_info=self._source_info(),
                trace_id=str(ULID()),
                reason_key=map_exception_to_reason(exc),
                error_message=str(exc),
                stack=traceback.format_exc(),
            )
            self._sink.add(trace)
            raise
        ended_at = _now_iso()
        latency_ms = int((time.monotonic() - started_monotonic) * 1000)
        usage = response.usage.model_dump() if getattr(response, "usage", None) else None
        trace = build_success_trace(
            request_snapshot=request_snapshot,
            response_usage=usage,
            response_tool_calls=None,
            identity=identity,
            timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
            env=self._env(),
            source_info=self._source_info(),
            trace_id=str(ULID()),
        )
        self._sink.add(trace)
        return response

    async def _invoke_non_streaming_async(
        self,
        *,
        inner_method: Any,
        kwargs: dict[str, Any],
    ) -> Any:
        per_call = kwargs.pop("autocontext", None)
        if kwargs.get("stream", False):
            raise NotImplementedError("async streaming wired in Task 2.8")
        identity = resolve_identity(per_call)
        request_snapshot = build_request_snapshot(
            model=kwargs.get("model", ""),
            messages=kwargs.get("messages", []),
            extra_kwargs={k: v for k, v in kwargs.items() if k not in {"model", "messages"}},
        )
        started_at = _now_iso()
        started_monotonic = time.monotonic()
        try:
            response = await inner_method(**kwargs)
        except Exception as exc:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            trace = build_failure_trace(
                request_snapshot=request_snapshot,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=self._env(),
                source_info=self._source_info(),
                trace_id=str(ULID()),
                reason_key=map_exception_to_reason(exc),
                error_message=str(exc),
                stack=traceback.format_exc(),
            )
            self._sink.add(trace)
            raise
        ended_at = _now_iso()
        latency_ms = int((time.monotonic() - started_monotonic) * 1000)
        usage = response.usage.model_dump() if getattr(response, "usage", None) else None
        tool_calls = None
        if hasattr(response, "choices") and response.choices and response.choices[0].message.tool_calls:
            tool_calls = [tc.model_dump() for tc in response.choices[0].message.tool_calls]
        trace = build_success_trace(
            request_snapshot=request_snapshot,
            response_usage=usage,
            response_tool_calls=tool_calls,
            identity=identity,
            timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
            env=self._env(),
            source_info=self._source_info(),
            trace_id=str(ULID()),
        )
        self._sink.add(trace)
        return response

    async def _invoke_non_streaming_async_responses(
        self,
        *,
        inner_method: Any,
        kwargs: dict[str, Any],
        normalized_messages: list[dict[str, Any]],
    ) -> Any:
        per_call = kwargs.pop("autocontext", None)
        identity = resolve_identity(per_call)
        model = kwargs.get("model", "")
        request_snapshot = build_request_snapshot(
            model=model,
            messages=normalized_messages,
            extra_kwargs={k: v for k, v in kwargs.items() if k not in {"model", "messages", "input"}},
        )
        started_at = _now_iso()
        started_monotonic = time.monotonic()
        try:
            response = await inner_method(**kwargs)
        except Exception as exc:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            trace = build_failure_trace(
                request_snapshot=request_snapshot,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=self._env(),
                source_info=self._source_info(),
                trace_id=str(ULID()),
                reason_key=map_exception_to_reason(exc),
                error_message=str(exc),
                stack=traceback.format_exc(),
            )
            self._sink.add(trace)
            raise
        ended_at = _now_iso()
        latency_ms = int((time.monotonic() - started_monotonic) * 1000)
        usage = response.usage.model_dump() if getattr(response, "usage", None) else None
        trace = build_success_trace(
            request_snapshot=request_snapshot,
            response_usage=usage,
            response_tool_calls=None,
            identity=identity,
            timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
            env=self._env(),
            source_info=self._source_info(),
            trace_id=str(ULID()),
        )
        self._sink.add(trace)
        return response

    def _invoke_streaming(
        self,
        *,
        inner_method: Any,
        kwargs: dict[str, Any],
    ) -> Any:
        per_call = kwargs.pop("autocontext", None)
        # Auto-inject stream_options.include_usage = True if absent.
        stream_opts = dict(kwargs.get("stream_options") or {})
        if "include_usage" not in stream_opts:
            stream_opts["include_usage"] = True
            kwargs["stream_options"] = stream_opts
        identity = resolve_identity(per_call)
        request_snapshot = build_request_snapshot(
            model=kwargs.get("model", ""),
            messages=kwargs.get("messages", []),
            extra_kwargs={k: v for k, v in kwargs.items() if k not in {"model", "messages"}},
        )
        started_at = _now_iso()
        started_monotonic = time.monotonic()

        inner_stream = inner_method(**kwargs)

        from autocontext.integrations.openai._stream import StreamProxy
        from autocontext.integrations.openai._trace_builder import finalize_streaming_trace

        # Store the accumulator reference in a dict so that the closure captures
        # the dict (not the proxy), avoiding a reference cycle that would prevent GC.
        acc_ref: dict[str, Any] = {"accumulator": None}
        sink = self._sink
        env = self._env()
        source_info = self._source_info()

        def on_finalize(outcome: dict[str, Any]) -> None:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            acc = acc_ref["accumulator"] or {"usage": None, "tool_calls": None}
            trace = finalize_streaming_trace(
                request_snapshot=request_snapshot,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=env,
                source_info=source_info,
                trace_id=str(ULID()),
                accumulated_usage=acc["usage"],
                accumulated_tool_calls=acc["tool_calls"],
                outcome=outcome,
            )
            sink.add(trace)

        proxy = StreamProxy(inner_stream=inner_stream, on_finalize=on_finalize)
        # Store the proxy's accumulator in acc_ref using a weakref to avoid a
        # cycle: proxy → on_finalize → acc_ref → proxy's accumulator
        # We link via the accumulator dict (not the proxy itself)
        acc_ref["accumulator"] = proxy._accumulator
        return proxy

    def _invoke_streaming_async(
        self,
        *,
        inner_method: Any,
        kwargs: dict[str, Any],
    ) -> Any:
        per_call = kwargs.pop("autocontext", None)
        # Auto-inject stream_options.include_usage = True if absent.
        stream_opts = dict(kwargs.get("stream_options") or {})
        if "include_usage" not in stream_opts:
            stream_opts["include_usage"] = True
            kwargs["stream_options"] = stream_opts
        identity = resolve_identity(per_call)
        request_snapshot = build_request_snapshot(
            model=kwargs.get("model", ""),
            messages=kwargs.get("messages", []),
            extra_kwargs={k: v for k, v in kwargs.items() if k not in {"model", "messages"}},
        )
        started_at = _now_iso()
        started_monotonic = time.monotonic()

        from autocontext.integrations.openai._stream import AsyncStreamProxy
        from autocontext.integrations.openai._trace_builder import finalize_streaming_trace

        acc_ref_async: dict[str, Any] = {"accumulator": None}
        sink = self._sink
        env = self._env()
        source_info = self._source_info()

        def on_finalize(outcome: dict[str, Any]) -> None:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            acc = acc_ref_async["accumulator"] or {"usage": None, "tool_calls": None}
            trace = finalize_streaming_trace(
                request_snapshot=request_snapshot,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=env,
                source_info=source_info,
                trace_id=str(ULID()),
                accumulated_usage=acc["usage"],
                accumulated_tool_calls=acc["tool_calls"],
                outcome=outcome,
            )
            sink.add(trace)

        async def _make_proxy() -> AsyncStreamProxy:
            coro_or_stream = inner_method(**kwargs)
            # AsyncCompletions.create may be a coroutine or direct async context manager
            if inspect.iscoroutine(coro_or_stream):
                inner_stream = await coro_or_stream
            else:
                inner_stream = coro_or_stream
            # If the stream itself is an async context manager, enter it
            if hasattr(inner_stream, "__aenter__"):
                inner_stream = await inner_stream.__aenter__()
            proxy = AsyncStreamProxy(inner_stream=inner_stream, on_finalize=on_finalize)
            # Link accumulator into acc_ref_async to avoid cycle
            acc_ref_async["accumulator"] = proxy._accumulator
            return proxy

        return _make_proxy()
